integration-tests/src/integration_test_circuits.rs

```diff
@@ -0,0 +1,332 @@
+        >(
+            verifier_params,
+            verifying_key,
+            strategy,
+            &[instances],
+            &mut verifier_transcript,
+        )
+        .expect("failed to verify bench circuit");
+    }
+
+    let general_params = get_general_params(degree);
+    let verifier_params: ParamsVerifierKZG<Bn256> = general_params.verifier_params().clone();
+
+    let verifying_key = keygen_vk(&general_params, &circuit).expect("keygen_vk should not fail");
+    let proving_key =
+        keygen_pk(&general_params, verifying_key, &circuit).expect("keygen_pk should not fail");
+
+    let transcript = Blake2bWrite::<_, G1Affine, Challenge255<_>>::init(vec![]);
+
+    // change instace to slice
+    let instance: Vec<&[Fr]> = instance.iter().map(|v| v.as_slice()).collect();
+
+    let proof = test_gen_proof(
+        RNG.clone(),
+        circuit,
+        &general_params,
+        &proving_key,
+        transcript,
+        &instance,
+    );
+
+    let verifying_key = proving_key.get_vk();
+    test_verify(
+        &general_params,
+        &verifier_params,
+        verifying_key,
+        &proof,
+        &instance,
+    );
+}
+
+fn test_mock<C: Circuit<Fr>>(degree: u32, circuit: &C, instance: Vec<Vec<Fr>>) {
+    let mock_prover = MockProver::<Fr>::run(degree, circuit, instance).unwrap();
+    mock_prover
+        .verify_par()
+        .expect("mock prover verification failed");
+}
+
+/// Integration test for evm circuit.
+pub async fn test_evm_circuit_block(block_num: u64, actual: bool) {
+    log::info!("test evm circuit, block number: {}", block_num);
+    let (builder, _) = gen_inputs(block_num).await;
+
+    let block = block_convert(&builder.block, &builder.code_db);
+
+    let degree = get_test_degree(&block);
+    let instance = get_test_instance(&block);
+    let circuit = get_test_cicuit_from_block(block);
+
+    test_mock(degree, &circuit, instance.clone());
+
+    if actual {
+        test_actual(degree, circuit, instance);
+    }
+}
+
+/// Integration test for state circuit.
+pub async fn test_state_circuit_block(block_num: u64, actual: bool) {
+    log::info!("test state circuit, block number: {}", block_num);
+
+    const DEGREE: u32 = 17;
+
+    let (builder, _) = gen_inputs(block_num).await;
+
+    // Generate state proof
+    // log via trace some of the container ops for debugging purposes
+    let stack_ops = builder.block.container.sorted_stack();
+    trace!("stack_ops: {:#?}", stack_ops);
+    let memory_ops = builder.block.container.sorted_memory();
+    trace!("memory_ops: {:#?}", memory_ops);
+    let storage_ops = builder.block.container.sorted_storage();
+    trace!("storage_ops: {:#?}", storage_ops);
+
+    // TODO: use rw_map from block
+    // let rw_map = RwMap::from(&builder.block.container);
+    let rw_map = RwMap::default();
+
+    let circuit = StateCircuit::<Fr>::new(rw_map, 1 << 16);
+    let instance = circuit.instance();
+
+    test_mock(DEGREE, &circuit, instance.clone());
```

Comment:
I think it would be better if all `test_mock` are only run when `!actual`.
Like: `if actual { test_actual } else { test_mock }`.

This is because the test_mock for big circuits (even with `verify_par`) can take some time.